File: Source/Engine/Graphics/D3D12/D3D12Sampler.cpp
// Copyright © Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

#include "D3D12Sampler.h"
#include "D3D12Graphics.h"

namespace Alimer
{
    namespace
    {
        [[nodiscard]] constexpr D3D12_FILTER_TYPE ToD3D12FilterType(SamplerFilter value)
        {
            switch (value)
            {
                case SamplerFilter::Nearest:
                    return D3D12_FILTER_TYPE_POINT;
                case SamplerFilter::Linear:
                    return D3D12_FILTER_TYPE_LINEAR;
                default:
                    ALIMER_UNREACHABLE();
                    return D3D12_FILTER_TYPE_POINT;
            }
        }

        [[nodiscard]] constexpr D3D12_TEXTURE_ADDRESS_MODE ToD3D12AddressMode(SamplerAddressMode mode)
        {
            switch (mode) {
                case SamplerAddressMode::Wrap:
                    return D3D12_TEXTURE_ADDRESS_MODE_WRAP;
                case SamplerAddressMode::Mirror:
                    return D3D12_TEXTURE_ADDRESS_MODE_MIRROR;
                case SamplerAddressMode::Clamp:
                    return D3D12_TEXTURE_ADDRESS_MODE_CLAMP;
                case SamplerAddressMode::Border:
                    return D3D12_TEXTURE_ADDRESS_MODE_BORDER;
                case SamplerAddressMode::MirrorOnce:
                    return D3D12_TEXTURE_ADDRESS_MODE_MIRROR_ONCE;
                default:
                    ALIMER_UNREACHABLE();
                    return D3D12_TEXTURE_ADDRESS_MODE_WRAP;
            }
        }
    }

	D3D12Sampler::D3D12Sampler(D3D12Graphics& device, const SamplerCreateInfo* info)
		: Sampler()
		, device{ device }
	{
        const D3D12_FILTER_TYPE minFilter = ToD3D12FilterType(info->minFilter);
        const D3D12_FILTER_TYPE magFilter = ToD3D12FilterType(info->magFilter);
        const D3D12_FILTER_TYPE mipFilter = ToD3D12FilterType(info->mipFilter);

        D3D12_FILTER_REDUCTION_TYPE reduction = info->compareFunction == CompareFunction::Never
            ? D3D12_FILTER_REDUCTION_TYPE_STANDARD
            : D3D12_FILTER_REDUCTION_TYPE_COMPARISON;

        D3D12_SAMPLER_DESC desc{};

        // https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_sampler_desc
        if (info->maxAnisotropy > 1)
        {
            desc.Filter = D3D12_ENCODE_ANISOTROPIC_FILTER(reduction);
        }
        else
        {
            desc.Filter = D3D12_ENCODE_BASIC_FILTER(minFilter, magFilter, mipFilter, reduction);
        }

        desc.AddressU = ToD3D12AddressMode(info->addressModeU);
        desc.AddressV = ToD3D12AddressMode(info->addressModeV);
        desc.AddressW = ToD3D12AddressMode(info->addressModeW);
        desc.MipLODBias = 0.f;
        desc.MaxAnisotropy = Min<UINT>(info->maxAnisotropy, 16u);
        if (info->compareFunction != CompareFunction::Never)
        {
            desc.ComparisonFunc = ToD3D12ComparisonFunc(info->compareFunction);
        }
        else
        {
            desc.ComparisonFunc = D3D12_COMPARISON_FUNC_NEVER;
        }

        desc.MinLOD = info->lodMinClamp;
        desc.MaxLOD = info->lodMaxClamp;
	}

	D3D12Sampler::~D3D12Sampler()
	{
		Destroy();
	}

	void D3D12Sampler::Destroy()
	{
	}
}

File: src/Alimer.Graphics/D3D12/D3D12Sampler.cs
// Copyright © Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

using static TerraFX.Interop.DirectX.DirectX;
using static TerraFX.Interop.DirectX.D3D12_FILTER_REDUCTION_TYPE;
using static TerraFX.Interop.DirectX.D3D12_COMPARISON_FUNC;
using static TerraFX.Interop.DirectX.D3D12_DESCRIPTOR_HEAP_TYPE;
using TerraFX.Interop.DirectX;

namespace Alimer.Graphics.D3D12;

internal sealed unsafe class D3D12Sampler : Sampler
{
    private readonly D3D12GraphicsDevice _device;
    private readonly D3D12_SAMPLER_DESC _desc;

    public D3D12Sampler(D3D12GraphicsDevice device, in SamplerDescription description)
        : base(description)
    {
        _device = device;
        D3D12_FILTER_TYPE minFilter = description.MinFilter.ToD3D12();
        D3D12_FILTER_TYPE magFilter = description.MagFilter.ToD3D12();
        D3D12_FILTER_TYPE mipmapFilter = description.MipFilter.ToD3D12();

        D3D12_FILTER_REDUCTION_TYPE reductionType = description.ReductionType.ToD3D12();

        _desc = new();

        // https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_sampler_desc
        _desc.MaxAnisotropy = Math.Min(Math.Max(description.MaxAnisotropy, 1u), 16u);
        if (_desc.MaxAnisotropy > 1)
        {
            _desc.Filter = D3D12_ENCODE_ANISOTROPIC_FILTER(reductionType);
        }
        else
        {
            _desc.Filter = D3D12_ENCODE_BASIC_FILTER(minFilter, magFilter, mipmapFilter, reductionType);
        }

        _desc.AddressU = description.AddressModeU.ToD3D12();
        _desc.AddressV = description.AddressModeV.ToD3D12();
        _desc.AddressW = description.AddressModeW.ToD3D12();
        _desc.MipLODBias = 0.0f;
        if (description.Compare != CompareFunction.Never)
        {
            _desc.ComparisonFunc = description.Compare.ToD3D12();
        }
        else
        {
            // Still set the function so it's not garbage.
            _desc.ComparisonFunc = D3D12_COMPARISON_FUNC_NEVER;
        }
        _desc.MinLOD = description.LodMinClamp;
        _desc.MaxLOD = description.LodMaxClamp;
    }

    /// <inheritdoc />
    public override GraphicsDevice Device => _device;

    public void CreateDescriptor(D3D12_CPU_DESCRIPTOR_HANDLE handle)
    {
        D3D12_SAMPLER_DESC desc = _desc;
        _device.Handle->CreateSampler(&desc, handle);
    }

    protected internal override void Destroy()
    {
    }
}
